source/sio/io_uring/async_accept.hpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2023 Xiaoming Zhang
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "sio/concepts.hpp"
+#include "sio/sequence/sequence_concepts.hpp"
+#include "sio/io_uring/file_handle.hpp"
+#include "sio/net/ip/address.hpp"
+
+#include <exec/env.hpp>
+#include <exec/linux/io_uring_context.hpp>
+#include <exec/trampoline_scheduler.hpp>
+#include <sys/socket.h>
+
+namespace sio {
+  namespace accept_ {
+    using namespace stdexec;
+
+    // TODO: with async_allocator
+    // using sender_t = async_allocator(sender_t);
+    using sender_t = io_uring::accept_sender;
+
+    template <class Receiver>
+    struct operation;
+
+    template <class Receiver>
+    struct next_receiver {
+      using is_receiver = void;
+
+      operation<Receiver>* op_;
+
+      void set_value(set_value_t) && noexcept {
+        op_->start_next();
+      }
+
+      void set_stopped(set_stopped_t) && noexcept {
+        if constexpr (unstoppable_token<stop_token_of_t<env_of_t<Receiver>>>) {
+          stdexec::set_value(static_cast<Receiver&&>(op_->rcvr_));
+        } else {
+          auto token = stdexec::get_stop_token(stdexec::get_env(op_->rcvr_));
+          if (token.stop_requested()) {
+            stdexec::set_stopped(static_cast<Receiver&&>(op_->rcvr_));
+          } else {
+            stdexec::set_value(static_cast<Receiver&&>(op_->rcvr_));
+          }
+        }
+      }
+
+      env_of_t<Receiver> get_env(get_env_t) const noexcept {
+        return stdexec::get_env(op_->rcvr_);
+      }
+    };
+
+    template <class Receiver>
+    struct operation {
+      Receiver rcvr_;
+      io_uring::acceptor& acceptor_;
+
+      using item_sender = decltype(stdexec::on(
+        std::declval<exec::trampoline_scheduler&>(),
+        std::declval<sender_t>()));
+
+      // !!!! The commented lines cause compilation errors.
+      // std::optional< connect_result_t<
+      //   exec::__next_sender_of_t<Receiver, item_sender>, //
+      //   next_receiver<Receiver>>>
+      //   op_{};
+      exec::trampoline_scheduler scheduler_{};
+
+      void start_next() noexcept {
+        // try {
+        //   stdexec::start(op_.emplace(__conv{[&] {
+        //     return stdexec::connect(
+        //       exec::set_next(
```

Comment:
Idk yet why this doesn't compile but we also want to close the socket after the next sender completes. Therefore we need something like this

```cpp
stdexec::connect(
  exec::finally(
       exec::set_next(rcvr_, sender_t{/* ... */} 
                             | stdexec::then([this](auto handle) { start_next(); return handle })),
       close_sender{/* ... */}),
  next_receiver<Receiver>{this})
```

Reply:
got it!